sail/src/main/java/org/apache/rya/rdftriplestore/inference/InferenceEngine.java

```diff
@@ -583,4 +625,69 @@ public boolean isSchedule() {
     public void setSchedule(boolean schedule) {
         this.schedule = schedule;
     }
+
+    /**
+     * For a given type, return any properties and values such that owl:hasValue restrictions on
+     * those properties could imply this type. No matter how many restrictions are returned, each
+     * one is considered individually sufficient: if a resource has the property and the value, then
+     * it belongs to the provided type. Takes type hierarchy into account, so the value may imply a
```

Comment:
If A were a superclass of B and C, rather than a subclass, then we would consider that individual to be of type A. This method is going to return all the hasValue restrictions on all of a type's subclasses, and since belonging to one subclass is sufficient to determine membership in the superclass, meeting one of those restrictions is also sufficient to determine membership in the superclass. Also keep in mind that hasValue property restrictions don't *require* any data to be there; they actually imply it: "x is a member of B <-> x has prop1=val1". So in your example, if an individual is known to be of type A, then it must also have types B and C, and it must also have prop1=val1 and prop2=val2. If those triples aren't in the data, that's OK, we can infer them (though that would be the other rewriting case and handled by the other method). The clearest  statement of this in the standards is probably [here](https://www.w3.org/TR/owl2-profiles/#Reasoning_in_OWL_2_RL_and_RDF_Graphs_using_Rules), specifically cls-hv1 and cls-hv2 in table 6. (This list of rules is incomplete since it only covers OWL RL languages, but the terms it does cover have their standard meaning.)